gtsam/slam/tests/testEssentialMatrixFactor.cpp

```diff
@@ -351,7 +349,227 @@ TEST (EssentialMatrixFactor3, minimization) {
+  Values initial;
+  EssentialMatrix initialE =
+      trueE.retract((Vector(5) << 0.1, -0.1, 0.1, 0.1, -0.1).finished());
+  initial.insert(1, initialE);
+  initial.insert(2, trueK);
+
+  // add prior factor for calibration
+  Vector5 priorNoiseModelSigma;
+  priorNoiseModelSigma << 10, 10, 10, 10, 10;
+  graph.emplace_shared<PriorFactor<Cal3_S2>>(
+      2, trueK, noiseModel::Diagonal::Sigmas(priorNoiseModelSigma));
+
+  LevenbergMarquardtOptimizer optimizer(graph, initial);
+  Values result = optimizer.optimize();
+
+  // Check result
+  EssentialMatrix actualE = result.at<EssentialMatrix>(1);
+  Cal3_S2 actualK = result.at<Cal3_S2>(2);
+  EXPECT(assert_equal(trueE, actualE, 1e-1));
+  EXPECT(assert_equal(trueK, actualK, 1e-2));
+
+  // Check error at result
+  EXPECT_DOUBLES_EQUAL(0, graph.error(result), 1e-4);
+
+  // Check errors individually
+  for (size_t i = 0; i < 5; i++)
+    EXPECT_DOUBLES_EQUAL(
+        0,
+        actualE.error(EssentialMatrix::Homogeneous(actualK.calibrate(pA(i))),
+                      EssentialMatrix::Homogeneous(actualK.calibrate(pB(i)))),
+        1e-6);
+}
+
+//*************************************************************************
+TEST(EssentialMatrixFactor4, minimizationWithWeakCal3S2Prior) {
+  // We need 7 points here as the prior on the focal length parameters is weak
+  // and the initialization is noisy. So in total we are trying to optimize 7
+  // DOF, with a strong prior on the remaining 3 DOF.
+  NonlinearFactorGraph graph;
+  for (size_t i = 0; i < 7; i++)
+    graph.emplace_shared<EssentialMatrixFactor4<Cal3_S2>>(1, 2, pA(i), pB(i),
+                                                          model1);
+
+  // Check error at ground truth
+  Values truth;
+  truth.insert(1, trueE);
+  truth.insert(2, trueK);
+  EXPECT_DOUBLES_EQUAL(0, graph.error(truth), 1e-8);
+
+  // Initialization
+  Values initial;
+  EssentialMatrix initialE =
+      trueE.retract((Vector(5) << 0.1, -0.1, 0.1, 0.1, -0.1).finished());
+  Cal3_S2 initialK =
+      trueK.retract((Vector(5) << 0.1, -0.1, 0.0, -0.0, 0.0).finished());
+  initial.insert(1, initialE);
+  initial.insert(2, initialK);
+
+  // add prior factor for calibration
+  Vector5 priorNoiseModelSigma;
+  priorNoiseModelSigma << 20, 20, 1, 1, 1;
+  graph.emplace_shared<PriorFactor<Cal3_S2>>(
+      2, initialK, noiseModel::Diagonal::Sigmas(priorNoiseModelSigma));
+
+  LevenbergMarquardtOptimizer optimizer(graph, initial);
+  Values result = optimizer.optimize();
+
+  // Check result
+  EssentialMatrix actualE = result.at<EssentialMatrix>(1);
+  Cal3_S2 actualK = result.at<Cal3_S2>(2);
+  EXPECT(assert_equal(trueE, actualE, 1e-1));
+  EXPECT(assert_equal(trueK, actualK, 1e-2));
+
+  // Check error at result
+  EXPECT_DOUBLES_EQUAL(0, graph.error(result), 1e-4);
+
+  // Check errors individually
+  for (size_t i = 0; i < 7; i++)
+    EXPECT_DOUBLES_EQUAL(
+        0,
+        actualE.error(EssentialMatrix::Homogeneous(actualK.calibrate(pA(i))),
```

Comment:
maybe add `using EssentialMatrix::Homogeneous` at top of file

Reply:
Does that work for class names? I get an `error: ‘gtsam::EssentialMatrix’ is not a namespace or unscoped enum`.